File: Gui/Linn/Gui/Resources/MessageTranslator.cs
using Linn.Gui;
using System.Xml;
using System.Collections.Generic;

namespace Linn {
namespace Gui {
namespace Resources {
    
public sealed class Translator : ISerialiseObject
{
    public Translator() {
    }
    
    public Translator(Message aFromMessage, Message aToMessage) {
        iFromMessage = aFromMessage;
        iToMessage = aToMessage;
    }
    
    public void Load(XmlNode aXmlNode) {
        XmlNodeList list;
        list = aXmlNode.SelectNodes("FromMessage");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iFromMessage = MessageFactory.Instance.Create(list[0]);
        }
        
        list = aXmlNode.SelectNodes("ToMessage");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iToMessage = MessageFactory.Instance.Create(list[0]);
        }
    }
    
    public void Link() {
        //Trace.WriteLine(Trace.kGui, "Linking Translator");
        //Trace.WriteLine(Trace.kGui, "iFromMessage=" + iFromMessage + ", iToMessage=" + iToMessage);
        iFromMessage.Link();
        iToMessage.Link();
    }
    
    public void Save(XmlTextWriter aWriter) {
        aWriter.WriteStartElement("FromMessage");
        string type = iFromMessage.GetType().ToString();
        aWriter.WriteAttributeString("class", type.Substring(type.LastIndexOf('.') + 1));
        iFromMessage.Save(aWriter);
        aWriter.WriteEndElement();

        aWriter.WriteStartElement("ToMessage");
        type = iToMessage.GetType().ToString();
        aWriter.WriteAttributeString("class", type.Substring(type.LastIndexOf('.') + 1));
        iToMessage.Save(aWriter);
        aWriter.WriteEndElement();
    }
    
    public Message ToMessage {
        get {
            return iToMessage;
        }
        set {
            iToMessage = value;
        }
    }
    
    public Message FromMessage {
        get {
            return iFromMessage;
        }
        set {
            iFromMessage = value;
        }
    }
    
    public Message Translate(Message aMessage) {
        if(aMessage.IsEqualTo(iFromMessage)) {
            return iToMessage;
        }
        return null;
    }
    
    private Message iToMessage = null;
    private Message iFromMessage = null;
}

    
public sealed class MessageTranslator : ISerialiseObject
{
    public void Load(XmlNode aXmlNode) {
        XmlNodeList list;
        list = aXmlNode.SelectNodes("Translator");
        if(list != null) {
            foreach(XmlNode node in list) {
                Translator translator = new Translator();
                translator.Load(node);
                iTranslatorList.Add(translator);
            }
        }
    }
    
    public void Link() {
        //Trace.WriteLine(Trace.kGui, "Linking MessageTranslator table");
        foreach(Translator t in iTranslatorList) {
            t.Link();
        }
    }
    
    public void Save(XmlTextWriter aWriter) {
        foreach(Translator translator in iTranslatorList) {
            aWriter.WriteStartElement("Translator");
            translator.Save(aWriter);
            aWriter.WriteEndElement();
        }
    }
    
    public void AddTranslator(Translator aTranslator) {
        iTranslatorList.Add(aTranslator);
    }
    
    public void RemoveTranslator(Translator aTranslator) {
        iTranslatorList.Remove(aTranslator);
    }
    
    public List<Translator> Translators {
        get {
            return iTranslatorList;
        }
    }
    
    public bool SendMessage(Message aMessage) {
        int numTranslated = 0;
        foreach(Translator translator in iTranslatorList) {
            Message message = translator.Translate(aMessage);
            if(message != null) {
                Trace.WriteLine(Trace.kGui, "MessageTranslator.SendMessage: aMessage(" + aMessage.Fullname + ")=" + aMessage + " to message(" + message.Fullname + ")=" + message);
                Messenger.Instance.PresentationMessage(message);
                numTranslated++;
            }
        }
        return(numTranslated > 0);
    }
    
    private List<Translator> iTranslatorList = new List<Translator>();
}
    
} // Resources
} // Gui
} // Linn

File: Gui/Linn/Gui/Resources/MessageTranslator.py
import Linn.Gui.Serialize
import Linn.Gui.Messenger
import Message

class Translator(Linn.Gui.Serialize.ISerialize):
    def __init__(self, aFromMsg=None, aToMsg=None):
        self.iFromMsg = aFromMsg
        self.iToMsg = aToMsg
        
    def Load(self, aStream):
        fromMsg = aStream.LoadNode('FromMessage')
        if len(fromMsg) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<FromMessage>', len(fromMsg), 1)
        
        classType = fromMsg[0].attributes[(None,'class')].value
        aStream.PushContextNode(fromMsg[0])
        self.iFromMsg = Message.CreateMessage(classType, aStream)
        aStream.PopContextNode()
        
        toMsg = aStream.LoadNode('ToMessage')
        if len(toMsg) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<ToMessage>', len(toMsg), 1)
        
        classType = toMsg[0].attributes[(None,'class')].value
        aStream.PushContextNode(toMsg[0])
        self.iToMsg = Message.CreateMessage(classType, aStream)
        aStream.PopContextNode()
    
    def Save(self, aStream):
        fromMsg = aStream.CreateElement('FromMessage')
        classType = aStream.CreateAttribute('class', None)
        fromMsg.setAttributeNodeNS(classType)
        aStream.PushContextNode(fromMsg)
        
        self.iFromMsg.Save(aStream)
        
        aStream.PopContextNode()
        aStream.SaveNode(fromMsg)
        
        toMsg = aStream.CreateElement('ToMessage')
        classType = aStream.CreateAttribute('class', None)
        toMsg.setAttributeNodeNS(classType)
        aStream.PushContextNode(toMsg)
        
        self.iToMsg.Save(aStream)
        
        aStream.PopContextNode()
        aStream.SaveNode(toMsg)
    
    def Link(self):
        pass
        
    def FromMsg(self):
        return self.iFromMsg
    
    def SetFromMsg(self, aFromMsg):
        self.iFromMsg = aFromMsg
    
    def ToMsg(self):
        return self.iToMsg
    
    def SetToMsg(self, aToMsg):
        self.iToMsg = aToMsg
        
    def Translate(self, aMessage):
        if aMessage == self.iFromMsg:
            #print 'Translating:\n', self.iFromMsg, 'to\n', self.iToMsg
            return self.iToMsg
        else:
            return None


class MessageTranslator(Linn.Gui.Serialize.ISerialize):
    def __init__(self):
        self.iTranslators = []
        
    def Load(self, aStream):
        translationList = aStream.LoadNode('Translator')
        for translation in translationList:
            translator = Translator()
            aStream.PushContextNode(translation)
            translator.Load(aStream)
            aStream.PopContextNode()
            self.iTranslators.append(translator)
    
    def Save(self, aStream):      
        for trans in self.iTranslators:
            translator = aStream.CreateElement('Translator')
            aStream.PushContextNode(translator)
            trans.Save(aStream)
            aStream.PopContextNode()
            aStream.SaveNode(translator)
    
    def Link(self):
        pass
        
    def AddTranslator(self, aFromMsg, aToMsg):
        self.iTranslators.append(Translator(aFromMsg, aToMsg))
        
    def RemoveTranslator(self, aFromMsg, aToMsg):
        for trans in self.iTranslators:
            if trans.FromMsg() == aFromMsg and trans.ToMsg() == aToMsg:
                self.iTranslators.remove(trans)
                
    def RemoveTranslatorAtIndex(self, aIndex):
        self.iTranslators.remove(self.iTranslators[aIndex])
                
    def TranslatorList(self):
        return self.iTranslators
        
    def SendMessage(self, aMsg):
        translated = False
        for trans in self.iTranslators:
            msg = trans.Translate(aMsg)
            if msg:
                Linn.Gui.Messenger.GetMessenger().PresentationMessage(msg)
                translated = True
        return translated
        
        
        